### PR TITLE
feat: Add gas optimized transaction deliver

### PR DESCRIPTION
**Pull request by koonopek**
Tested on goerli, works stable even with gas spiking 5 times. Here is the contract I was writing to:
https://goerli.etherscan.io/address/0xddededae617cec1b1c7b60c467ac6b9787047b9f

In default setup we prefer time of delivery over low cost

### CODE REVIEW

File: packages/rpc-providers/src/TransactionDeliver.ts [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+type FeeStructure = {
+  maxFeePerGas: number;
+  maxPriorityFeePerGas: number;
+};
+
+const unsafeBnToNumber = (bn: BigNumber) => Number(bn.toString());
+
+const ethGasTrackerOracle: GasOracleFn = async (apiKey: string = "") => {
+  const response = // rate limit is 5 seconds
```
Comment (hatskier) [inline packages/rpc-providers/src/TransactionDeliver.ts:59]: Won't we hit the rate limit many times then?

Comment (hatskier) [reply packages/rpc-providers/src/TransactionDeliver.ts:59]: Shouldn't we implement caching here?

File: packages/rpc-providers/src/TransactionDeliver.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+};
+
+type FeeStructure = {
+  maxFeePerGas: number;
+  maxPriorityFeePerGas: number;
+};
+
+const unsafeBnToNumber = (bn: BigNumber) => Number(bn.toString());
+
+const ethGasTrackerOracle: GasOracleFn = async (apiKey: string = "") => {
```
Comment (hatskier) [inline packages/rpc-providers/src/TransactionDeliver.ts:58]: Can you name function as a verb please?

File: packages/rpc-providers/src/TransactionDeliver.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+} as Record<number, GasOracleFn | undefined>;
+
+const DEFAULT_TRANSACTION_DELIVERY_OPTS = {
+  maxAttempts: 5,
+  priorityFeePerGasMultiplier: 1.125, // 112,5%
+  percentileOfPriorityFee: 75,
+  logger: (text: string) => console.log(`[${TransactionDeliver.name}] ${text}`),
+};
+
+export class TransactionDeliver {
```
Comment (hatskier) [inline packages/rpc-providers/src/TransactionDeliver.ts:91]: Maybe TransactionDeliveryMan?

File: packages/rpc-providers/src/TransactionDeliver.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  public async deliver<T extends Contract, M extends keyof T>(
+    contract: T,
+    method: M,
+    params: Parameters<T[M]>,
+    gasLimit?: number
+  ): Promise<TransactionResponse | undefined> {
+    const provider = contract.provider as providers.JsonRpcProvider;
+    const address = await contract.signer.getAddress();
+
+    let lastAttempt = undefined as LastDeliveryAttempt | undefined;
```
Comment (hatskier) [inline packages/rpc-providers/src/TransactionDeliver.ts:107]: I would use `let lastAttempt: LastDeliveryAttempt | undefined = undefined;` according to typescript best practices

File: packages/rpc-providers/src/TransactionDeliver.ts (hunk 2 above)
Comment (koonopek) [reply packages/rpc-providers/src/TransactionDeliver.ts:58]: Yep

File: packages/rpc-providers/src/TransactionDeliver.ts (hunk 3 above)
Comment (koonopek) [reply packages/rpc-providers/src/TransactionDeliver.ts:91]: Yes! I was thinking about it ;)

File: packages/rpc-providers/src/TransactionDeliver.ts (hunk 4 above)
Comment (koonopek) [reply packages/rpc-providers/src/TransactionDeliver.ts:107]: Good point :)

File: packages/rpc-providers/src/TransactionDeliver.ts (hunk 1 above)
Comment (koonopek) [reply packages/rpc-providers/src/TransactionDeliver.ts:59]: Good idea :) 5s caching should solve all problems

Comment (koonopek) [reply packages/rpc-providers/src/TransactionDeliver.ts:59]: But it still wouldn't hit the rate limit cause we call this function once every 15 seconds there is no point IN ETH blockchain to call it more frequently

File: packages/on-chain-relayer/src/core/contract-interactions/update-prices.ts [hunk 5]
```diff
@@ -67,10 +72,22 @@ const updatePricesInPriceFeedsAdapter = async ({
...
+  const wrappedContract = wrapContract(adapterContract);
+
+  const deliveryResult = await deliveryMan.deliver(
+    wrappedContract,
+    "updateDataFeedsValues",
+    [proposedTimestamp],
+    TX_CONFIG.gasLimit ? Number(TX_CONFIG.gasLimit) : undefined
   );
+
+  if (!deliveryResult) {
```
[the remark this answers is not in the document]
Comment (koonopek) [reply packages/on-chain-relayer/src/core/contract-interactions/update-prices.ts:84]: Yeah good point :)

**Review by hatskier — APPROVED** [review]

**Pull request merged**